### ISSUE TITLE
Add `react/no-object-type-as-default-prop` to `react/recommended`

### ISSUE THREAD

**Issue by SimonSimCity** [jsx-eslint/eslint-plugin-react#3863]
I propose to add the rule `react/no-object-type-as-default-prop` at least with `warning` if not `error` to the `react/recommended` set.

Reason: I don't see a good reason not to enable it. It helps you not to fall into the trap of endless rendering. I even had a case where I wrote a component which just seemed to fail in my test environment, but just because this was the only place I didn't set the prop which had an object as default.

If you see a reason not to enable this, I'm thrilled to hear!

**Comment by SimonSimCity** [comment]
Here's just one way of creating a component which will have an endless render:

```tsx
export const MyObject = ({
  myFancyThing = { test: 1 }, // Never set an object here as default! It will be a different object every time this element is rendered!
}: {
  myFancyThing?: { test: number };
}) => {
  const [foo, setFoo] = useState<string | undefined>(undefined);

  useEffect(() => {
    setFoo("")
  }, [myFancyThing]); // This will always run, because the dependency is always a different object!

  return (
    <...>
  );
};
```

**Comment by ljharb** [comment]
Adding anything to the recommended config is a breaking change, but I agree it should be done at some point.
